game/hud/src/components/HUD/HUDEditor.tsx

```diff
@@ -0,0 +1,463 @@
+                  onClick={ () => this.resetWidget(this.props.selectedWidget.name) }
+                ></i>
+              </a>
+            </HUDEditorToolbarItem>
+
+          </HUDEditorToolbar>
+        </div> }
+      </HUDEditorContainer>
+    );
+  }
+
+  public componentDidMount() {
+    window.addEventListener('mouseup', this.onMouseUp);
+    window.addEventListener('mousemove', this.onMouseMove);
+    this.props.setSelectedWidget(null);
+  }
+
+  public componentDidUnMount() {
+    window.removeEventListener('mouseup', this.onMouseUp);
+    window.removeEventListener('mousemove', this.onMouseMove);
+  }
+
+  private onMouseDown = (e: any, mode: EditMode) => {
+    // check if we can do this or not...
+
+    // for now we always allow it
+    this.setMode(mode);
+    e.preventDefault();
+    e.stopPropagation();
+    this.lastPosition = { x: e.screenX , y: e.screenY };
+  }
+
+  private onMouseUp = () => {
+    this.lastPosition = { x: NaN , y: NaN };
+    this.mouseDownForScaleHold = false;
+    if (this.mouseScaleHoldInitTimeout != null) {
+      clearTimeout(this.mouseScaleHoldInitTimeout);
+    }
+    this.mouseDownForOpacityHold = false;
+    if (this.mouseOpacityHoldInitTimeout != null) {
+      clearTimeout(this.mouseOpacityHoldInitTimeout);
+    }
+    if (this.state.mode !== EditMode.NONE) this.setMode(EditMode.NONE);
+  }
+
+  private mouseMovement = (e: MouseEvent) => {
+    if (this.lastPosition.x === NaN) {
+      // just starting to move, so 0 move
+      this.lastPosition = { x: e.screenX, y: e.screenY };
+      return { x: 0, y: 0 };
+    }
+    const movement = { x: e.screenX - this.lastPosition.x, y: e.screenY - this.lastPosition.y };
+    this.lastPosition = { x: e.screenX, y: e.screenY };
+    return movement;
+  }
+
+  private onMouseMove = (e: MouseEvent | any) => {
+    const name = this.props.selectedWidget ? this.props.selectedWidget.name : null;
+    const widget = this.props.selectedWidget ? this.props.selectedWidget.widget : null;
+    const mouseMove = this.mouseMovement(e);
+    switch (this.state.mode) {
+      default: return;
+      case EditMode.NONE: return;
+      case EditMode.SCALE:
+        {
+          this.setScale(
+            name,
+            widget,
+            this.props.selectedWidget.widget.position.scale + (mouseMove.y * this.state.scaleFactor * -1),
+          );
+        }
+        break;
+      case EditMode.OPACITY:
+        {
+          this.setOpacity(name, widget, this.props.selectedWidget.widget.position.opacity + (mouseMove.y * -0.01));
+        }
+        break;
+      case EditMode.MOVEEDITOR:
+        {
+          this.setState({
+            editorPosition: {
+              x: this.state.editorPosition.x - mouseMove.x,
+              y: this.state.editorPosition.y + mouseMove.y,
+            },
+          });
+        }
+        break;
+    }
+  }
+
+  private setMode = (m: EditMode) => {
+    this.setState({
+      mode: m,
+    } as any);
```

Comment:
I think this has been here for a while, but we don't need this `as any`

Reply:
✔️